### ISSUE TITLE
Accessing the memory of a Dataset created as "diskless"

### ISSUE THREAD

**Issue by nmassey001** [Unidata/netcdf4-python#865]
When creating a Dataset with "diskless=True":

`ncd = netCDF4.Dataset(filepath, mode='w', diskless=True)`

It doesn't seem possible to access the buffer for the Dataset.

I've looked in the source code and the call to the underlying C library is:
` ...
                elif diskless:
                    if persist:
                        ierr = nc_create(path, NC_WRITE | NC_CLOBBER |
                                NC_DISKLESS | NC_PERSIST, &grpid)
                    else:
                        ierr = nc_create(path, NC_CLOBBER | NC_DISKLESS , &grpid)
...
`

The Python library doesn't seem to record a reference to the buffer created by nc_create.
Would it be possible to be able to expose this buffer?

**Comment by jswhit** [comment]
As is stands now you can create a netcdf file from an existing python memory buffer, but you can't create a reference to the memory of an existing in-memory Dataset.  The former is done by using the INMEMORY capability of the C lib, the latter using the DISKLESS capability, which are separate (https://www.unidata.ucar.edu/software/netcdf/docs_rc/md__Users_wfisher_Desktop_gitprojects_netcdf-c_docs_inmemory.html).  However, from reading that page again  it does look like it may be possible to re-access the buffer of a INMEMORY file after data has been written to it using the `nc_close_memio` function.

**Comment by jswhit** [comment]
Just curious, what's your use-case?

**Comment by jswhit** [comment]
@DennisHeimbigner - is it possible to access the memory buffer of a file created with the NC_DISKLESS flag?  Or, is it preferable to create an in-memory file using `nc_create_mem` and then access the memory buffer from `nc_close_memio`?

**Comment by DennisHeimbigner** [comment]
A case I had not considered. Looking at the code, it appears that you must
use nc_create_mem in order to extract the memory content.

**Comment by jswhit** [comment]
Right now, `netCDF4.Dataset` creates a (read-only) in-memory file if` mode='r'` and a memory buffer is passed in via the `memory` kwarg.  This could be extended to `mode='w'`, and a memory buffer could be returned when the `close` method is invoked.  In this case, I guess the `memory` kwarg would just be used to indicate an in-memory file is desired (if it is not equal to `None`).  Would this satisfy your use case?

**Comment by DennisHeimbigner** [comment]
You probably need either another close operation or a keyword for existing close
to indicate that you want the in-memory data.

**Comment by jswhit** [comment]
Pull request #866 adds this functionality.  I've tested it with the following script

```python
from netCDF4 import Dataset
import numpy as np
format = 'NETCDF4_CLASSIC'
nc = Dataset('test.nc','w',memory=1,format=format)
nc.history = 'test of nc_create_mem/nc_close_memio'
d = nc.createDimension('x',None)
v = nc.createVariable('v',np.int32,'x')
v[0:5] = np.arange(5)
f = open('test.nc','wb')
f.write(nc.close())
f.close()
```

**Comment by DennisHeimbigner** [comment]
How is the memory returned by nc_close_memio accessed?

**Comment by jswhit** [comment]
`Dataset.close` returns a python `bytes` object created from the memory buffer returned by `nc_close_memio`. You really can't do much with it in python except write it to a file or open it as a Dataset again.

**Comment by jswhit** [comment]
Instead of returning a `bytes` object, `Dataset.close()` now returns a custom cython memory buffer (which implements the python buffer protocol).  This avoids making an extra copy of the memory.

**Comment by jswhit** [comment]
That's good.  I'd like to be sure this implementation doesn't leak before I merge into master.  Do you think you could devise a test to check for memory leakage?

**Comment by nmassey001** [comment]
Here's a more thorough test, by creating arrays of various sizes, in loops and cleaning up the memory with `del`.  I've used psutil to monitor the memory usage of the python Process.
I'd normally use valgrind to check for memory leaks on exit, but it's not available yet on Mac OS Mojave!

[nc4_memio_test.txt](https://github.com/Unidata/netcdf4-python/files/2808134/nc4_memio_test.txt)

And the results are as follow:

```
==== Test 1 - smaller than available memory, no cleanup ====
	 start: 26980352
	 end  : 1143816192
	 diff : 1116835840
	 array: 1113600000

==== Test 2 - smaller than available memory, with cleanup ====
	 start: 1143816192
	 end  : 1144410112
	 diff : 593920
	 array: 1113600000

==== Test 3 - smaller than available memory, in a loop with cleanup ====
	 Loop: 0	 diff: 147456
	 Loop: 1	 diff: 0
	 Loop: 2	 diff: 73728
	 Loop: 3	 diff: 12288
	 Loop: 4	 diff: 12288
	 Loop: 5	 diff: 20480
	 Loop: 6	 diff: 0
	 Loop: 7	 diff: 16384
	 Loop: 8	 diff: 0
	 Loop: 9	 diff: 0

==== Test 4 - large array, with cleanup ====
	 start: 31039488
	 end  : 13176832
	 diff : -17862656
	 array: 11136000000
```

**Comment by nmassey001** [comment]
I think that shows that there are not any memory leaks.
It would be worth noting to the user that they are responsible for cleaning up the buffer (using del), at least within the python program.

Thanks, Neil

**Comment by jswhit** [comment]
What happens if you don't use del?

**Comment by nmassey001** [comment]
The first example doesn't use del - the difference in the start mem and end mem is roughly the size of the array.
I'd like to check what happens when you then exit python, but I can't check with valgrind.  Should the garbage collector collect this?

**Comment by jswhit** [comment]
All del does is decrement the reference count, the memory is only released when the reference count gets to zero.  If you use the same name ('bytebuffer') for all the tests, then the reference count should go to zero and the memory should be reclaimed without del.

**Comment by nmassey001** [comment]
Here are the results from using `bytebuffer` as the variable name all the way through:

```
==== Test 1 - smaller than available memory, no cleanup ====
	 start: 27074560
	 end  : 1143861248
	 diff : 1116786688
	 array: 1113600000

==== Test 2 - smaller than available memory, no cleanup ====
	 start: 1143861248
	 end  : 1144397824
	 diff : 536576
	 array: 1113600000

==== Test 3 - smaller than available memory, in a loop no cleanup ====
	 Loop: 0	 diff: 69632
	 Loop: 1	 diff: 16384
	 Loop: 2	 diff: 12288
	 Loop: 3	 diff: 24576
	 Loop: 4	 diff: 12288
	 Loop: 5	 diff: 32768
	 Loop: 6	 diff: 69632
	 Loop: 7	 diff: 40960
	 Loop: 8	 diff: 8192
	 Loop: 9	 diff: 24576

==== Test 4 - large array, no cleanup ====
	 start: 1144709120
	 end  : 1371119616
	 diff : 226410496
	 array: 5568000000
```

**Comment by nmassey001** [comment]
It does seem to be.  I've just run the test with the loop again and get:
```
	 Loop: 0	 diff: 1116708864
	 Loop: 1	 diff: 626688
	 Loop: 2	 diff: 106496
	 Loop: 3	 diff: 8192
	 Loop: 4	 diff: 16384
	 Loop: 5	 diff: 73728
	 Loop: 6	 diff: 0
	 Loop: 7	 diff: 0
	 Loop: 8	 diff: 16384
	 Loop: 9	 diff: 8192
```
(Ignore the first one as that's the allocation of the array).
So, it's not consistent, either, as there are some iterations where the memory usage didn't increase.

**Comment by jswhit** [comment]
I think there is some variability in when the memory is actually released.  If I run your test and watch the memory usage in top it oscillates up and down, but does not show an upward trend.

**Comment by DennisHeimbigner** [comment]
There is a detail that may be relevant. From the inmemory.md documentation:
> 2. If the *NC_MEMIO_LOCKED* flag is <b>not</b> set, then
> the netcdf library will take control of the incoming memory.
> This means that the user should not make any attempt to free
> or even read the incoming memory block in this case

I do not know if this is a problem for python.

**Comment by jswhit** [comment]
Seems like NC_MEMIO_LOCKED is only relevant if nc_open_mem is called (and a block of memory is provided).    Right now the python interface is not setting this flag, so the C lib is allowed to re-allocate or free the memory.   However, the way I've set up the python interface, nc_open_mem is only called with mode='r' (that is, if you can only create a read-only Dataset from a memory buffer).  This means that nc_close_memio is never called, so I think that NC_MEMIO_LOCKED not being set should not be an issue.

nc_close_memio is only called from the python interface if mode='w', and the Dataset is created with nc_create_mem (so no memory buffer is provided).

**Comment by jswhit** [comment]
@DennisHeimbigner  - this reminds me that I had a question about the `initialsize` parameter in `nc_create_mem`.  Am I right in assuming that this value is only used for NETCDF3 files, and essentially ignored by HDF5?  For NETCDF3, what happens if the value of `initialsize` given is too small?

**Comment by DennisHeimbigner** [comment]
Since that parameter is for create, it does not matter if it is too small. 
The size will be realloc'd larger as needed.

**Comment by DennisHeimbigner** [comment]
My comment above about NC_MEMIO_LOCKED flag applies when it is NOT set,
which is your case, I think.

**Comment by jswhit** [comment]
Yes, that's right.  I don't think it's a problem for us since we are not using nc_open_memio.

**Comment by jswhit** [comment]
Pull request merged, closing now.  @nmassey001, if you have any issues with this feature, please reopen this ticket.
